fix(editor-web-component): Render modal inside shadow root

CodingDive opened the pull request:
Migrated to radix/react-dialog. React-modal was really outdated and the `appElement` was killing our integrations. 

Here are the docs. https://www.radix-ui.com/primitives/docs/components/dialog  Behavior and styles should stay the same.

Commit:
fix(editor-web-component): Render modal inside shadow root

Commit:
Merge branch 'staging' into fix/editor-web-component-modal-in-shadow-root

Commit:
Merge branch 'staging' into fix/editor-web-component-modal-in-shadow-root

Commit:
wip(editor-web-component): Trying to hack react modal to work within shadow root

Commit:
refactor(modal): Migrate from react-modal to Radix dialog

Commit:
Merge branch 'staging' into fix/editor-web-component-modal-in-shadow-root

Commit:
fix(editor-web-component): Try to fix use blur on outside click (shadow root is null)

Commit:
fix(editor-web-component): Modal inside shadow root

Commit:
fix(editor-web-component): Render modal in all plugins within shadow root

Commit:
chore(editor): Release v0.8.2 of editor

Commit:
chore(prettier): Run prettier

elbotho commented on packages/editor/src/plugins/audio/toolbar.tsx:
```diff
@@ -26,6 +31,8 @@ export const AudioToolbar = ({
   setShowSettingsModal: Dispatch<SetStateAction<boolean>>
 }) => {
   const audioStrings = useEditorStrings().plugins.audio
+  const ref = useRef<HTMLButtonElement>(null)
+  const shadowRoot = useShadowRoot(ref)
```
I don't understand why you pass the opener-button ref as the containerRef to `useShadowRoot`. Could you explain?
Also I'd love to move this complexity inside the modal instead of having it in all the consumers. Would that be possible somehow?

CodingDive replied:
We just need any ref to detect whether it's inside the shadow root or not. 

>Also I'd love to move this complexity inside the modal instead of having it in all the consumers. Would that be possible somehow?

Maybe rendering an empty div inside the modal with a ref. That could be enough! Will try!

CodingDive replied:
Worked great! Thank you for the suggestion :heart:

elbotho replied:
Nice, happy about that!

Commit:
refactor(editor): Remove complexity, handle div within modal component

elbotho commented on packages/editor/src/core/helpers/use-shadow-root.ts:
```diff
@@ -22,3 +22,19 @@ export function useShadowRoot(containerRef: React.RefObject<HTMLElement>) {
+  return shadowRoot instanceof ShadowRoot
+}
+
+export function getFirstElementOrUndefined(
+  shadowRoot: ShadowRoot | null | undefined
+): HTMLElement | undefined {
+  if (!shadowRoot) {
+    return undefined
+  }
+
+  return shadowRoot?.firstElementChild as HTMLElement
+}
```
possible simplifications:

```suggestion
export function getFirstElementOrUndefined(shadowRoot?: ShadowRoot | null) {
  return shadowRoot?.firstElementChild as HTMLElement | undefined
}
```

CodingDive replied:
I like the way I wrote it better. Some more lines, but more clarity and easier types

elbotho commented on packages/editor/src/core/hooks/use-blur-on-outside-click.ts:
```diff
@@ -22,7 +22,7 @@ export function useBlurOnOutsideClick(
         root.contains(clickedElement) && // clicked element is present in the document
         editorWrapperRef.current && // provided wrapper is defined
         !editorWrapperRef.current.contains(clickedElement) && // clicked element is not a child of the provided wrapper
-        !clickedElement.closest('.ReactModalPortal') // clicked element is not a part of a modal
+        !clickedElement.closest('.serlo-modal') // clicked element is not a part of a modal
```
nice that you thought of those

elbotho reviewed (commented):
did not test yet, but code looks good!

elbotho commented:
Some changes in behaviour: 

**Share modal**
When I open the share modal for  the first time the close button is focused / has an outline:

<img width="518" alt="image" src="https://github.com/serlo/frontend/assets/1258870/ee0da029-208a-4977-bf05-1c59bf09c52a">

----

**Editor setting modal** (e.g. image-plugin):
restoring/keeping focus works only when I use the close button. If I click outside the focus goes to document.body.
(works in staging)

----

**Editor Article Add Modal** ([example](https://frontend-git-fix-editor-web-component-modal-in-sha-e87021-serlo.vercel.app/entity/repository/add-revision/1565)):
Also looses focus to body on closing (close button and outside click)

<img width="873" alt="image" src="https://github.com/serlo/frontend/assets/1258870/5f094f47-dcb6-45af-b209-40a78e56cbf0">

Commit:
fix(modal): Implement feedback of Botho. Focus first focusable element that is not the close button and restore focus to previous element upon close

CodingDive commented:
Thank you for the thorough review @elbotho 

> When I open the share modal for the first time the close button is focused / has an outline

According to aria recommendations, one should focus the first focusable element. That should clearly not be the close button though! Radix was just picking the first element, so I moved the button down.

https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/

Now when I share, the URL gets focused.

![image](https://github.com/serlo/frontend/assets/28842311/07671dd2-7197-4560-9980-39fa5886a1d8)


> Editor Article Add Modal ([example](https://frontend-git-fix-editor-web-component-modal-in-sha-e87021-serlo.vercel.app/entity/repository/add-revision/1565)):
> Also looses focus to body on closing (close button and outside click)

Should be upon ESC/closing via button, but the outside click does not yet work. 

> restoring/keeping focus works only when I use the close button. If I click outside the focus goes to document.body.
(works in staging)

I believe the reason why this does not yet work, is that the toolbar looses focus through `useBlurOnOutsideClick` before the `useEffect` https://github.com/serlo/frontend/pull/3842/files#diff-1c5900ff0d9c5494c672ce3039511959d8a50fe827a74bf903a43c75ed06f787R49 I just added has time to execute. One risky fix I can think of is to restore focus on unmount, but I feel this could have a lot of unintended consequences.

Commit:
fix(modal): Restore previous state when clicking away, experiment with uncontrolled modal

Commit:
fix(modal): Get rid of trigger and uncontrolled component

* Refactor onCloseClick to setIsOpen to simplify code and keep outside components in sync with the internal state of radix
* Remove trigger and uncontrolled component of modal with close button code

Commit:
chore: Remove console log

CodingDive commented:
Please give it another round of testing! I hope I addressed all the issues

shn-srl commented on apps/web/src/components/user-tools/revision/checkout-reject-buttons.tsx:
```diff
@@ -62,7 +62,7 @@ export function CheckoutRejectButtons({
       <ModalWithCloseButton
         isOpen={modalMode !== null}
-        onCloseClick={onCloseClick}
+        setIsOpen={(open) => (open ? void null : onCloseClick())}
```
Maybe we can have setIsOpen as an optional prop? Then here we could have a cleaner prop,


`open && onCloseClick()`

CodingDive replied:
`setIsOpen` needs to be required for the state management to work. It's just that some consumers exclusively need the `onClose` handler. 
I find the above code more readable than saying

`setIsOpen={(open) => (!open && onCloseClick())}` but it'd save us the ternary :thinking:

shn-srl commented on apps/web/src/components/modal-with-close-button.tsx:
```diff
@@ -22,74 +24,120 @@ interface ModalWithCloseButtonProps {
+      // Need to add this tiny delay when unmounting the modal, so that the
+      // use-blur-on-outside-click has time to detect whether the modal is open
+      // or not.
+      if (open === false) {
```
Nit: Can we return early here too? 

```
if(open !== false){
 setIsOpen(open)
 return
}

...
```

shn-srl commented on apps/web/src/components/modal-with-close-button.tsx:
```diff
@@ -22,74 +24,120 @@ interface ModalWithCloseButtonProps {
+        previouslyFocusedElementRef.current.focus()
+      }
+    }
+  }, [isOpen, shadowRoot])
```
Nitpick: Return early would be nice here too, to reduce nesting avoid the need for the else of the outer scope.

Asked ChatGPT its opinion and it provided the code below, which is a bigger refactorm looks kinda nice but I haven't tested it

import { useEffect, useRef } from 'react';

const useFocusManagement = (isOpen, shadowRoot) => {
  const previouslyFocusedElementRef = useRef(null);

  useEffect(() => {
    const getActiveElement = () => (shadowRoot ? shadowRoot.activeElement : document.activeElement);

    if (isOpen) {
      previouslyFocusedElementRef.current = getActiveElement();
    } else if (previouslyFocusedElementRef.current) {
      previouslyFocusedElementRef.current.focus();
    }
  }, [isOpen, shadowRoot]);
};

export default useFocusManagement;

Commit:
Merge branch 'staging' into fix/editor-web-component-modal-in-shadow-root

Commit:
refactor(modal): Feedback of Shane, early exits

elbotho reviewed (approved):
I'm a bit sad that we have to keep track of the previous focus ourselves, but it seems to work alright. thank you.

CodingDive commented:
> I'm a bit sad that we have to keep track of the previous focus ourselves, but it seems to work alright. thank you.

Same. This works from the library with ESC already, but for the outside click, we have to manage ourselves :crying_cat_face:
